File: C++/CombSort.cpp
#include <iostream>
#include <stdlib.h>
#include <vector>

using namespace std;

void print_array(auto array)
{
    for(int i : array)
    {
        cout << i << " ";
    }
    cout << endl;
}

void bubble_sort(auto &array)
{
    auto len = array.size();
    int k = 0;
    for (int i = 0; i < len; i++)
    {
        bool flag = false;
        for (int j = len - 1; j > i; j--)
        {
            if (array[j] < array[j-1])
            {
                flag = true;
                auto temp = array[j-1];
                array[j-1] = array[j];
                array[j] = temp;
            }
        }
        
        if(not flag)
        {
            break;
        }

        cout << "after " << ++k << " change: ";
        print_array(array);
    }
}

void comb_sort(auto &array)
{
    auto len = array.size();
    auto step = len;
    auto k = 0;
    while((step/=1.3) > 1)
    {
        for (int i = 0; i < len - step; i++)
        {
            if(array[i] > array[i+step])
            {
                int temp = array[i];
                array[i] = array[i+step];
                array[i+step] = temp;
            }
        }

        cout << "after " << ++k << " change: ";
        print_array(array);
    }

    cout << "before bubble: ";
    print_array(array);

    bubble_sort(array);
}

int main()
{  
    vector<int> array;
    char ch;
    int num = 0;
    cout << "please input the array(split by space):";
    while((ch = cin.get()) != '\n')
    {
        if(ch != ' ')
        {
            num = num * 10 + int(ch - '0');
        }    
        else if(ch == ' ')
        {
            array.push_back(num);
            num = 0;
        }
            
    }

    array.push_back(num);

    cout << "origin is: " ;
    print_array(array);

    comb_sort(array);
    cout << "result is: " ;
    print_array(array);

    system("pause");
    return 0;
}

File: python/CombSort.py
import datetime

def bubble_sort_ascended(a):
    k =0
    for i in range(0, len(a)):
        flag = False
        k = k + 1
        for j in range(len(a)-1, i, -1):
            if(a[j] < a[j-1]):
                flag = True
                a[j], a[j-1] = a[j-1], a[j]
        if(not flag):
            break
        print("after %d change: "%k, a)
    return a

def comb_sort_ascended(a):
    step = len(a)
    t = 0
    while (step > 1):
        t = t + 1
        k = int(step)
        for i in range(0, len(a)-k):
            if(a[i] > a[i+k]):
                a[i], a[i+k] = a[i+k], a[i]
        step = step/1.3
        print("after %d comb change: "%t, a)
    
    print("before bubble: ", a)

    bubble_sort_ascended(a)
    return a


def bubble_sort_dascended(a):
    k =0
    for i in range(0, len(a)):
        flag = False
        k = k + 1
        for j in range(len(a)-1, i, -1):
            if(a[j] > a[j-1]):
                flag = True
                a[j], a[j-1] = a[j-1], a[j]
        if(not flag):
            break
        print("after %d change: "%k, a)
    return a

def comb_sort_dascended(a):
    step = len(a)
    t = 0
    while (step > 1):
        t = t + 1
        k = int(step)
        for i in range(0, len(a)-k):
            if(a[i] < a[i+k]):
                a[i], a[i+k] = a[i+k], a[i]
        step = step/1.3
        print("after %d comb change: "%t, a)
    
    print("before bubble: ", a)

    bubble_sort_dascended(a)
    return a

def main():
    start = datetime.datetime.now()
    Array = list(map(int,input("please input the array(split by space):").split()))
    print("input is: ", Array)
    SortedArrayAscend = comb_sort_ascended(Array)
    print("ascend result is: ", SortedArrayAscend)

    SortedArrayDascend = comb_sort_dascended(Array)
    print("dascend result is: ", SortedArrayDascend)
    end = datetime.datetime.now()
    print(end - start)

if __name__ == "__main__":
    main()